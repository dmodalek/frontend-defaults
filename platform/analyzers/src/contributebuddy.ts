import { directoryExists, getDependencyInstallation, DependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ContributebuddyAnalyzerResult = {
    contributebuddy: boolean;
    contributebuddyConfigFolder?: string,
    contributebuddyInstallation?: DependencyInstallation
};

export const contributebuddyAnalyzer = async (cwd: string): Promise<ContributebuddyAnalyzerResult> => {
    const CONTRIBUTE_BUDDY_FOLDER = join(cwd, '.contributebuddy');
    const contributebuddyConfigFolder = await directoryExists(CONTRIBUTE_BUDDY_FOLDER);
    const contributebuddyInstallation = await getDependencyInstallation(cwd, 'contribute-buddy');

    if (contributebuddyConfigFolder || contributebuddyInstallation.declared) {
        return {
            contributebuddy: true,
            contributebuddyConfigFolder: CONTRIBUTE_BUDDY_FOLDER,
            contributebuddyInstallation

        }
    }

    return {
        contributebuddy: false,
    };
}