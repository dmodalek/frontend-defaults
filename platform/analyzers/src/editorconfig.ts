import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type EditorconfigAnalyzerResult = {
	editorConfig: boolean;
};

export class EditorconfigAnalyzer extends Analyzer<EditorconfigAnalyzerResult> {
	async analyze(): Promise<EditorconfigAnalyzerResult> {
		const eidtorConfigExists = await fileExists(join(this.context, '.editorconfig'));

		return {
			editorConfig: eidtorConfigExists,
		};
	}
}
