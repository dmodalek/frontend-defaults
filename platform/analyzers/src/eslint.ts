import { DependencyInstallation, fileExists, getDependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ESLintAnalyzerResult = {
	eslint: boolean;
	eslintIgnore?: boolean;
	eslintInstallation?: DependencyInstallation
};

export const eslintAnalyzer = async (cwd: string): Promise<ESLintAnalyzerResult> => {
	const doesESLintExist = await fileExists(join(cwd, '.eslintrc.js'));

	if (doesESLintExist) {
		const doesESLintIgnoreExist = await fileExists(join(cwd, '.eslintignore'));
		const eslintInstallation = await getDependencyInstallation(cwd, 'eslint');

		return {
			eslint: doesESLintExist,
			eslintIgnore: doesESLintIgnoreExist,
			eslintInstallation
		};
	}

	return {
		eslint: doesESLintExist,
	};
}