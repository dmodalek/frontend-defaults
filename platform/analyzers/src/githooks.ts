import {
	DependencyInstallation,
	fileExists,
	getDependencyInstallation,
	getPackageJSON,
	IPackage,
} from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitHooksLocation = 'rc' | 'package';

export type GitHooksAnalyzerResult = {
	gitHooks: boolean;
	gitHooksInstallation?: DependencyInstallation;
	gitHooksLocation?: GitHooksLocation;
};

type PackageWithHuskyConfig = IPackage & {
	husky: any;
};

export const gitHooksAnalyzer = async (cwd: string): Promise<GitHooksAnalyzerResult> => {
	const doesGitHooksConfigExist = await fileExists(join(cwd, '.huskyrc'));
	const gitHooksInstallation = await getDependencyInstallation(cwd, 'husky');
	const pkg: PackageWithHuskyConfig = (await getPackageJSON(cwd)) as any;

	if (doesGitHooksConfigExist || pkg.husky) {
		return {
			gitHooks: doesGitHooksConfigExist,
			gitHooksLocation: pkg.husky ? 'package' : 'rc',
			gitHooksInstallation,
		};
	}

	return {
		gitHooks: false,
	};
};
