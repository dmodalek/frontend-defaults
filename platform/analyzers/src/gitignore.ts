import { fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitIgnoreAnalyzerResult = {
	gitignore: boolean;
};

export const gitIgnoreAnalyzer = async (cwd: string): Promise<GitIgnoreAnalyzerResult> => {
	const doesGitignoreExist = await fileExists(join(cwd, '.gitignore'));

	return {
		gitignore: doesGitignoreExist,
	};
}