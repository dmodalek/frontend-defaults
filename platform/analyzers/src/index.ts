// export all related analyzer stuff for contributebuddy
export { contributebuddyAnalyzer, ContributebuddyAnalyzerResult } from './contributebuddy';

// export all related analyzer stuff for editorconfig
export { editorConfigAnalyzer, EditorconfigAnalyzerResult } from './editorconfig';

// export all related analyzer stuff for eslint
export { eslintAnalyzer, ESLintAnalyzerResult } from './eslint';

// export all related analyzer stuff for githooks
export { gitHooksAnalyzer, GitHooksAnalyzerResult } from './githooks';

// export all related analyzer stuff for gitignore
export { gitIgnoreAnalyzer, GitIgnoreAnalyzerResult } from './gitignore';

// export all related analyzer stuff for license
export { licenseAnalyzer, LicenseAnalyzerResult } from './license';

// export all related analyzer stuff for nitro
export { nitroAnalyzer, NitroAnalyzerResult } from './nitro';

// export all related analyzer stuff for node
export { nodeAnalyzer, NodeAnalyzerResult } from './node';

// export all related analyzer stuff for npmrc
export { npmrcAnalyzer, NPMRCAnalyzerResult } from './npmrc';

// export all related analyzer stuff for prettier
export { prettierAnalyzer, PrettierAnalyzerResult } from './prettier';

// export all related analyzer stuff for readme
export { readmeAnalyzer, ReadmeAnalyzerResult } from './readme';

// export all related analyzer stuff for repository
export { repositoryAnalyzer, RepositoryAnalyzerResult } from './repository';

// export all related analyzer stuff for stylelint
export { stylelintAnalyzer, StylelintAnalyzerResult } from './stylelint';

// export all related analyzer stuff for tslint
export { tslintAnalyzer, TSLintAnalyzerResult } from './tslint';

// export all related analyzer stuff for typescript
export { typescriptAnalyzer, TypeScriptAnalyzerResult } from './typescript';

// export all related analyzer stuff for webpack
export { webpackAnalyzer, WebpackAnalyzerResult } from './webpack';
