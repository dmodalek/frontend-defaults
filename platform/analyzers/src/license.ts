import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type LicenseAnalyzerResult = {
	license: boolean;
};

export class LicenseAnalyzer extends Analyzer<LicenseAnalyzerResult> {
	async analyze(): Promise<LicenseAnalyzerResult> {
		const licenseExists = await fileExists(join(this.context, 'LICENSE'));

		return {
			license: licenseExists,
		};
	}
}
