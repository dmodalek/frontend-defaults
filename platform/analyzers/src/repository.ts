import { fileExists, getStats, getJSON, DependencyInstallation, getDependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type RepositoryAnalyzerResult = {
	size: {
		source: number,
		vendor: number
	},
	projectType: 'default' | 'angular' | 'react',
	lerna: boolean,
	lernaProjectInformation: LernaConfiguration | undefined,
	lernaInstallation: DependencyInstallation | undefined,
	nx: boolean,
	nxProjectInformation: NxConfiguration | undefined,
	nxInstallation: DependencyInstallation | undefined
};

export type LernaConfiguration = {
	packages: string[],
	version: string,
	npmClient?: string
}

type NxProject = {
	tags?: string[],
}

export type NxConfiguration = {
	npmScope: string,
	projects: {
		[name: string]: NxProject
	},
	implicitDependencies?: {
		[dependency: string]: string
	}
}

export const repositoryAnalyzer = async (cwd: string): Promise<RepositoryAnalyzerResult> => {
	const LERNA_PATH = join(cwd, 'lerna.json');
	const NX_PATH = join(cwd, 'nx.json');

	const lernaExists = await fileExists(LERNA_PATH);
	const nxExists = await fileExists(NX_PATH);
	const rootStats = await getStats(cwd) || ({ size: -1 } as any);
	const vendorStats = await getStats(join(cwd, 'node_modules')) || ({ size: -1 } as any);

	const reactInstallation = await getDependencyInstallation(cwd, 'react');
	const isReactProject = Boolean(reactInstallation.declared);
	const angularInstallation = await getDependencyInstallation(cwd, '@angular/core');
	const isAngularProject = Boolean(angularInstallation.declared);


	let lernaProjectInformation: LernaConfiguration | undefined;
	let lernaInstallation: DependencyInstallation | undefined;
	let nxProjectInformation: NxConfiguration | undefined;
	let nxInstallation: DependencyInstallation | undefined;

	if (lernaExists) {
		lernaProjectInformation = await getJSON<LernaConfiguration>(LERNA_PATH);
		lernaInstallation = await getDependencyInstallation(cwd, 'lerna');
	}

	if (nxExists) {
		nxProjectInformation = await getJSON<NxConfiguration>(NX_PATH);
		nxInstallation = await getDependencyInstallation(cwd, '@nrwl/nx');
	}

	return {
		size: {
			source: rootStats.size,
			vendor: vendorStats.size
		},
		projectType: isReactProject ? 'react' : (isAngularProject ? 'angular' : 'default'),
		lerna: lernaExists,
		lernaProjectInformation,
		lernaInstallation,
		nx: nxExists,
		nxProjectInformation,
		nxInstallation,
	};
}