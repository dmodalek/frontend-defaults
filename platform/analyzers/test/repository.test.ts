import { getFixtureContext } from './utils';
import { repositoryAnalyzer } from '../src/repository';

const FIXTURE_LERNA = getFixtureContext('monorepo-project-lerna');
const FIXTURE_LERNA_NI = getFixtureContext('monorepo-project-lerna-not-installed');
const FIXTURE_NX = getFixtureContext('monorepo-project-nrwl');
const FIXTURE_NX_NI = getFixtureContext('monorepo-project-nrwl-not-installed');
const FIXTURE_NONE = getFixtureContext('ts-project-not-installed'); // any fixture without monrepo

describe('Analyzers', () => {
    describe('RepositoryAnalyzer', () => {
        it('should not crash the ProjectAnalyzer', async () => {
            expect(async () => await repositoryAnalyzer(FIXTURE_LERNA)).not.toThrow();
            expect(async () => await repositoryAnalyzer(FIXTURE_LERNA_NI)).not.toThrow();
            expect(async () => await repositoryAnalyzer(FIXTURE_NX)).not.toThrow();
            expect(async () => await repositoryAnalyzer(FIXTURE_NX_NI)).not.toThrow();
            expect(async () => await repositoryAnalyzer(FIXTURE_NONE)).not.toThrow();
        });

        it('should analyze a non-monorepo-ish project correctly', async () => {
            const analytics = await repositoryAnalyzer(FIXTURE_NONE);

            expect(analytics.nx).toEqual(false);
            expect(analytics.lerna).toEqual(false);
            expect(analytics.nxInstallation).toBeUndefined();
            expect(analytics.lernaInstallation).toBeUndefined();
            expect(analytics.nxProjectInformation).toBeUndefined();
            expect(analytics.lernaProjectInformation).toBeUndefined();
        });

        it('should analyze a lerna project correctly', async () => {
            const analytics = await repositoryAnalyzer(FIXTURE_LERNA);

            expect(analytics.lerna).toEqual(true);
            expect(analytics.lernaInstallation!.declared).toEqual('3.0.0');
            expect(analytics.lernaInstallation!.installed).toEqual('3.13.1');
            expect(analytics.lernaInstallation).toMatchSnapshot({
                latest: expect.any(String) // don't save latest version in snapshot
            });
            expect(analytics.lernaProjectInformation).toMatchSnapshot();
        });

        it('should analyze a lerna project without installation correctly', async () => {
            const analytics = await repositoryAnalyzer(FIXTURE_LERNA_NI);

            expect(analytics.lerna).toEqual(true);
            expect(analytics.lernaInstallation!.declared).toEqual('3.0.0');
            expect(analytics.lernaInstallation!.installed).toBeNull();
            expect(analytics.lernaInstallation).toMatchSnapshot({
                latest: expect.any(String) // don't save latest version in snapshot
            });
        });

        it('should analyze a nrwl/nx project correctly', async () => {
            const analytics = await repositoryAnalyzer(FIXTURE_NX);

            expect(analytics.nx).toEqual(true);
            expect(analytics.nxInstallation!.declared).toEqual('6.2.0');
            expect(analytics.nxInstallation!.installed).toEqual('7.7.2');
            expect(analytics.nxInstallation).toMatchSnapshot({
                latest: expect.any(String) // don't save latest version in snapshot
            });
            expect(analytics.nxProjectInformation).toMatchSnapshot();
        });

        it('should analyze a nrwl/nx project without installation correctly', async () => {
            const analytics = await repositoryAnalyzer(FIXTURE_NX_NI);

            expect(analytics.nx).toEqual(true);
            expect(analytics.nxInstallation!.declared).toEqual('6.2.0');
            expect(analytics.nxInstallation!.installed).toBeNull();
            expect(analytics.nxInstallation).toMatchSnapshot({
                latest: expect.any(String) // don't save latest version in snapshot
            });
        });
    });
});
