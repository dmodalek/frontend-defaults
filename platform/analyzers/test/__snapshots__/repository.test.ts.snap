// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyzers RepositoryAnalyzer should analyze a lerna project correctly 1`] = `
Object {
  "declared": "3.0.0",
  "installed": "3.13.1",
  "latest": Any<String>,
  "location": "devDependencies",
}
`;

exports[`Analyzers RepositoryAnalyzer should analyze a lerna project correctly 2`] = `
Object {
  "packages": Array [
    "packages/*",
  ],
  "version": "3.0.0",
}
`;

exports[`Analyzers RepositoryAnalyzer should analyze a lerna project correctly 3`] = `
Object {
  "source": 192,
  "vendor": 96,
}
`;

exports[`Analyzers RepositoryAnalyzer should analyze a lerna project without installation correctly 1`] = `
Object {
  "declared": "3.0.0",
  "installed": null,
  "latest": Any<String>,
  "location": "devDependencies",
}
`;

exports[`Analyzers RepositoryAnalyzer should analyze a nrwl/nx project correctly 1`] = `
Object {
  "declared": "6.2.0",
  "installed": "7.7.2",
  "latest": Any<String>,
  "location": "devDependencies",
}
`;

exports[`Analyzers RepositoryAnalyzer should analyze a nrwl/nx project correctly 2`] = `
Object {
  "implicitDependencies": Object {
    "nx.json": "*",
  },
  "npmScope": "nx-examples",
  "projects": Object {
    "example": Object {
      "tags": Array [
        "test-suite",
      ],
    },
  },
}
`;

exports[`Analyzers RepositoryAnalyzer should analyze a nrwl/nx project correctly 3`] = `
Object {
  "source": 192,
  "vendor": 96,
}
`;

exports[`Analyzers RepositoryAnalyzer should analyze a nrwl/nx project without installation correctly 1`] = `
Object {
  "declared": "6.2.0",
  "installed": null,
  "latest": Any<String>,
  "location": "devDependencies",
}
`;
